src/contracts/libraries/MagnitudeCheckpoints.sol

```diff
@@ -0,0 +1,266 @@
+     * `low` and `high` define a section where to do the search, with inclusive `low` and exclusive `high`.
+     *
+     * WARNING: `high` should not be greater than the array's length.
+     */
+    function _lowerBinaryLookup(
+        MagnitudeCheckpoint[] storage self,
+        uint32 timestamp,
+        uint256 low,
+        uint256 high
+    ) private view returns (uint256) {
+        while (low < high) {
+            uint256 mid = MathUpgradeable.average(low, high);
+            if (_unsafeAccess(self, mid)._timestamp < timestamp) {
+                low = mid + 1;
+            } else {
+                high = mid;
+            }
+        }
+        return high;
+    }
+
+    /**
+     * @dev Access an element of the array without performing bounds check. The position is assumed to be within bounds.
+     */
+    function _unsafeAccess(
+        MagnitudeCheckpoint[] storage self,
+        uint256 pos
+    ) private pure returns (MagnitudeCheckpoint storage result) {
+        assembly {
+            mstore(0, self.slot)
+            result.slot := add(keccak256(0, 0x20), pos)
+        }
+    }
+
+    /**
+     *
+     *                         ADDITIONAL FUNCTIONS FROM EIGEN-LABS
+     *
+     */
+
+    /**
+     * @dev Returns the value in the last (most recent) checkpoint with timestamp lower or equal than the search timestamp, or zero if there is none.
+     * In addition, returns the position of the checkpoint in the array.
+     */
+    function upperLookupWithPos(History storage self, uint32 timestamp) internal view returns (uint224, uint256) {
+        uint256 len = self._magnitudes.length;
+        uint256 pos = _upperBinaryLookup(self._magnitudes, timestamp, 0, len);
+        return pos == 0 ? (0, pos) : (_unsafeAccess(self._magnitudes, pos - 1)._value, pos - 1);
+    }
+
+    /**
+     * @dev Returns the value in the last (most recent) checkpoint with timestamp lower or equal than the search timestamp, or zero if there is none.
+     * In addition, returns the position of the checkpoint in the array.
+     *
+     * NOTE: This is a variant of {upperLookup} that is optimised to find "recent" checkpoint (checkpoints with high timestamps).
+     */
+    function upperLookupRecentWithPos(History storage self, uint32 timestamp) internal view returns (uint224, uint256) {
+        uint256 len = self._magnitudes.length;
+
+        uint256 low = 0;
+        uint256 high = len;
+
+        if (len > 5) {
+            uint256 mid = len - MathUpgradeable.sqrt(len);
+            if (timestamp < _unsafeAccess(self._magnitudes, mid)._timestamp) {
+                high = mid;
+            } else {
+                low = mid + 1;
+            }
+        }
+
+        uint256 pos = _upperBinaryLookup(self._magnitudes, timestamp, low, high);
+
+        return pos == 0 ? (0, pos) : (_unsafeAccess(self._magnitudes, pos - 1)._value, pos - 1);
+    }
+
+    /// @notice WARNING: this function is only used because of the invariant property
```

Comment:
can a library inherit another library, would prefer that...

Reply:
The checkpoints upgradeable library has additional checkpoint structs and functions for different size keys, values pairs like below. In an attempt to save some additional bytecode bloat and to make the naming more clear I felt this was more convenient. Also wanted to add functions like `decrementCurrentAndFutureCheckpoints`.
Could be worth discussing

```
struct Trace160 {
      Checkpoint160[] _checkpoints;
  }

  struct Checkpoint160 {
      uint96 _key;
      uint160 _value;
  }
 ```